lib/auth/predicate/partial/solver.go

```diff
@@ -0,0 +1,406 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aclements/go-z3/z3"
+	"github.com/gravitational/trace"
+)
+
+// Resolver resolves an identifier to a value. Returns nil if the identifier is not found and is unknown.
+type Resolver func([]string) any
+
+// Type represents a type of a value within a predicate expression.
+type Type int
+
+const (
+	// TypeInt is an Go int type.
+	TypeInt Type = iota
+
+	// TypeBool is a Go bool type.
+	TypeBool
+
+	// TypeString is a Go string type.
+	TypeString
+)
+
+// Solver is a solver instance which can be used to solve partial predicate expressions.
+// This is rather expensive to create so it should preferably be reused when possible.
+type Solver struct {
+	def    *z3.Context
+	solver *z3.Solver
+}
+
+// NewSolver creates a new solver instance.
+func NewSolver() *Solver {
+	config := z3.NewContextConfig()
+	def := z3.NewContext(config)
+	solver := z3.NewSolver(def)
+	return &Solver{def, solver}
+}
+
+// PartialSolveForAll solves a given partial predicate expression for all possible values of the given identifier.
+// The expression must have a boolean output type and the type of the unknown identifier must be specified.
+// On timeout, no solutions are returned.
+func (s *Solver) PartialSolveForAll(predicate string, resolveIdentifier Resolver, querying string, to Type, timeout time.Duration) ([]z3.Value, error) {
+	outCh := make(chan []z3.Value, 1)
+	errCh := make(chan error, 1)
+
+	// Run the solver in a separate goroutine to allow the caller to handle timeouts.
+	go func() {
+		defer close(outCh)
+		defer close(errCh)
+
+		out, err := s.partialSolveForAllImpl(predicate, resolveIdentifier, querying, to)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		outCh <- out
+	}()
+
+	select {
+	case out := <-outCh:
+		return out, nil
+	case err := <-errCh:
+		return nil, err
+	case <-time.After(timeout):
+		s.def.Interrupt()
+		// wait for the other goroutine to cleanup
+		<-outCh
+		return nil, trace.LimitExceeded("timeout")
+	}
+}
+
+// partialSolveForAllImpl is the implementation of PartialSolveForAll that runs logic directly.
+func (s *Solver) partialSolveForAllImpl(predicate string, resolveIdentifier Resolver, querying string, to Type) ([]z3.Value, error) {
+	// parse the predicate expression into a Go AST
+	ast, err := parser.ParseExpr(predicate)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a context struct which is used to share various information obtained during lowering
+	ctx := &ctx{s.def, s.solver, make(map[string]z3.Value), z3.Sort{}, resolveIdentifier}
+
+	// reset the state of the Z3 solver on exit so that it can be reused
+	defer ctx.solver.Reset()
+
+	// create a Z3 identifier for the unknown value
+	switch to {
+	case TypeInt:
+		ctx.unkTy = s.def.IntSort()
+	case TypeBool:
+		ctx.unkTy = s.def.BoolSort()
+	case TypeString:
+		ctx.unkTy = s.def.StringSort()
+	default:
+		return nil, trace.BadParameter("unsupported output type %v", to)
+	}
+
+	// lower the Go AST into a Z3 AST
```

Comment:
what does it mean to "lower" an AST? That's what I'm not very clear on and was hoping for a bit of info

Reply:
Hi, sorry. This is a term we take for granted within interpreter and compiler dev circles. Please see the additional information provided on the lower function.